### ISSUE TITLE
String.split/toUpperCase/toLowerCase/replaceAll do not work

### ISSUE THREAD

**Issue by objectProgrammer** [mirkosertic/Bytecoder#266]
EDIT: I found that this same error also affects toUpperCase and toLowerCase.
EDIT: I found that a similar error affects replaceAll

```java
String string = "A;b;c";
String[] strings = string.split(";");
for (String s : strings) Console.console().log(s);
Console.console().log(string.toUpperCase());
Console.console().log(string.toLowerCase());
```

2019-11-04 compile-time error:
```
Caused by: java.lang.IllegalStateException: Don't know how to promote BOOLEAN to REFERENCE
        at de.mirkosertic.bytecoder.ssa.TypeRef$Native$4.eventuallyPromoteTo(TypeRef.java:91)
        at de.mirkosertic.bytecoder.ssa.Value.widestTypeOf(Value.java:103)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.newPHIFor(ParsingHelperCache.java:203)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.lambda$resolveInitialPHIStateForNode$3(ParsingHelperCache.java:161)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.getLocalVariable(ParsingHelper.java:104)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.requestValue(ParsingHelper.java:120)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.newPHIFor(ParsingHelperCache.java:192)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.lambda$resolveInitialPHIStateForNode$3(ParsingHelperCache.java:161)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.getLocalVariable(ParsingHelper.java:104)
        at de.mirkosertic.bytecoder.ssa.NaiveProgramGenerator.initializeBlockWith(NaiveProgramGenerator.java:675)
        at de.mirkosertic.bytecoder.ssa.NaiveProgramGenerator.initializeBlock(NaiveProgramGenerator.java:379)
        at de.mirkosertic.bytecoder.ssa.NaiveProgramGenerator.generateFrom(NaiveProgramGenerator.java:268)
```

```java
String string = "Hello, Earth";
Console.console().log(string.replaceAll("(Hello, )\\w+", "$1Mars"));
```

```
Caused by: java.lang.IllegalStateException: Don't know how to promote REFERENCE to BOOLEAN
        at de.mirkosertic.bytecoder.ssa.TypeRef$Native$9.eventuallyPromoteTo(TypeRef.java:170)
        at de.mirkosertic.bytecoder.ssa.Value.widestTypeOf(Value.java:103)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.newPHIFor(ParsingHelperCache.java:203)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.lambda$resolveInitialPHIStateForNode$1(ParsingHelperCache.java:128)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.getLocalVariable(ParsingHelper.java:104)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.requestValue(ParsingHelper.java:120)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.getLocalVariable(ParsingHelper.java:104)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.requestValue(ParsingHelper.java:120)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.getLocalVariable(ParsingHelper.java:104)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.requestValue(ParsingHelper.java:120)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.getLocalVariable(ParsingHelper.java:104)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.requestValue(ParsingHelper.java:120)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.getLocalVariable(ParsingHelper.java:104)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.requestValue(ParsingHelper.java:120)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.newPHIFor(ParsingHelperCache.java:192)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.lambda$resolveInitialPHIStateForNode$3(ParsingHelperCache.java:161)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.getLocalVariable(ParsingHelper.java:104)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.requestValue(ParsingHelper.java:120)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.newPHIFor(ParsingHelperCache.java:192)
        at de.mirkosertic.bytecoder.ssa.ParsingHelperCache.lambda$resolveInitialPHIStateForNode$3(ParsingHelperCache.java:161)
        at de.mirkosertic.bytecoder.ssa.ParsingHelper.getLocalVariable(ParsingHelper.java:104)
        at de.mirkosertic.bytecoder.ssa.NaiveProgramGenerator.initializeBlockWith(NaiveProgramGenerator.java:675)
        at de.mirkosertic.bytecoder.ssa.NaiveProgramGenerator.initializeBlock(NaiveProgramGenerator.java:379)
        at de.mirkosertic.bytecoder.ssa.NaiveProgramGenerator.generateFrom(NaiveProgramGenerator.java:268)
```

<details>
<summary>(Click here) Leftover warnings from a debug?</summary>
<pre>
Class.forName usage detected with unknown class name
Class sun.nio.cs.UTF_8 is used by reflection!
Class sun.nio.cs.UTF_8 is used by reflection!
Class sun.nio.cs.UTF_16 is used by reflection!
Class sun.nio.cs.ISO_8859_1 is used by reflection!
Class sun.nio.cs.UTF_8 is used by reflection!
Class sun.nio.cs.UTF_16 is used by reflection!
Class sun.nio.cs.ISO_8859_1 is used by reflection!
Class java.lang.CharacterDataLatin1 is used by reflection!
Class java.lang.CharacterData00 is used by reflection!
Class java.lang.CharacterData01 is used by reflection!
Class java.lang.CharacterData02 is used by reflection!
Class java.lang.CharacterData0E is used by reflection!
Class java.lang.CharacterDataPrivateUse is used by reflection!
Class java.lang.CharacterDataUndefined is used by reflection!
</pre>
</details>

**Comment by mirkosertic** [comment]
I guess this is the root cause, and the wrapping exception message was
`de.mirkosertic.bytecoder.ssa.ControlFlowProcessingException: Error processing CFG for sun/util/locale/provider/LocaleProviderAdapter.forType`, right?

Its an exception handler with PHI values that cannot be resolved correctly. I thought I fixed that many moons ago, but here is another corner case I missed. It is damn hard to get JVM Exception handling right.

**Comment by mirkosertic** [comment]
About the Reflection API usages in the compiler log: is is not a warning or an error. It is meant to inform the user about certain Reflection API usages. The used classes are not automatically resolved, as some cases are not always needed, as seen by Latin1 Character encodings vs, whole UTF vs. generated code size.

**Comment by objectProgrammer** [comment]
Yes indeed that was the message, I thought it was unnecessary so I left it out. How about providing an option to disable to Reflection API usage logger? It's pretty unnecessary to me.

**Comment by mirkosertic** [comment]
Merged to master. All other RegEx related issues will be resolved by #269.
